### PR TITLE
core: make current*Block atomic, and accessor functions mutex-free

### PR DESCRIPTION
**Pull request by holiman**
Supersedes https://github.com/ethereum/go-ethereum/pull/15980 . 
This PR makes accesses to `CurrentBlock()`, `CurrentFastBlock()`  mutex-free, using atomics instead. Previously, the lower p2p-levels sometimes were blocked in their communication with peers because they need e.g. difficulty or block numbers. With this PR, that should not be a problem. 

The use of atomic does _not_ change the fact that mutexes still need to be held when setting the values, so that hasn't been changed. 

I'm currently resuming a fast-sync with this code, to see that there are no issues with e.g. switching over pivot or stuff like that.

### CODE REVIEW

**Comment by karalabe** [comment]
Could you do the same for currentHeader inside headerchain.go?

File: core/blockchain.go [hunk 1]
```diff
@@ -239,18 +239,20 @@ func (bc *BlockChain) loadLastState() error {
 	bc.currentFastBlock = bc.currentBlock
 	if head := GetHeadFastBlockHash(bc.db); head != (common.Hash{}) {
 		if block := bc.GetBlockByHash(head); block != nil {
-			bc.currentFastBlock = block
+			bc.currentFastBlock.Store(block)
 		}
 	}
 
 	// Issue a status log for the user
+	currentFastBlock := bc.currentFastBlock.Load().(*types.Block)
```
Comment (karalabe) [inline core/blockchain.go]: Could `bc.currentFastBlock.Load().(*types.Block)` be replaced with `bc.CurrentFastBlock()`?

Seems you're doing it below, wondering why not here.

File: core/blockchain.go [hunk 2]
```diff
@@ -330,28 +334,19 @@ func (bc *BlockChain) FastSyncCommitHead(hash common.Hash) error {
 
 // GasLimit returns the gas limit of the current HEAD block.
 func (bc *BlockChain) GasLimit() uint64 {
-	bc.mu.RLock()
-	defer bc.mu.RUnlock()
-
-	return bc.currentBlock.GasLimit()
+	return bc.currentBlock.Load().(*types.Block).GasLimit()
```
Comment (karalabe) [inline core/blockchain.go]: `bc.currentBlock.Load().(*types.Block)` -> `bc.CurrentBlock()`?

**Comment by holiman** [comment]
Regarding doing the same for headerchain... This is a snippet of it: 
```golang
func (hc *HeaderChain) CurrentHeader() *types.Header {
	return hc.currentHeader
}
```
There are no mutexes there. Not even for setting: 
```golang
// SetCurrentHeader sets the current head header of the canonical chain.
func (hc *HeaderChain) SetCurrentHeader(head *types.Header) {
	if err := WriteHeadHeaderHash(hc.chainDb, head.Hash()); err != nil {
		log.Crit("Failed to insert head header hash", "err", err)
	}
	hc.currentHeader = head
	hc.currentHeaderHash = head.Hash()
}
```
It looks very dangerous, but I'm no expert on header-chain... But anyhow, there's no point adding there.

**Comment by holiman** [comment]
Fixed, ptal

**Comment by karalabe** [comment]
The reason there aren't any mutexes is because the headerchain is not directly accessible, rather it's embedded into core.blockchain and light.lightchain. However, both instances have locks protecting the access, so they face the same congestion.

**Comment by karalabe** [comment]
e.g.

```

// CurrentHeader retrieves the current head header of the canonical chain. The
// header is retrieved from the HeaderChain's internal cache.
func (bc *BlockChain) CurrentHeader() *types.Header {
	bc.mu.RLock()
	defer bc.mu.RUnlock()

	return bc.hc.CurrentHeader()
}
```

**Comment by karalabe** [comment]
It's a very fragile design btw, the reason for handling the locking in blockchain/lightchain was due to the blockchain needing to be able to lock both current block and current header at the same time, which is not possible if there are two sets of locks.

**Comment by holiman** [comment]
If so, should both currentHeader and `currentHeaderHash` be fixed up, or only the former?

**Comment by karalabe** [comment]
I don't think `currentHeaderHash` is accessed from the outside, it seems to just be a cache since headers don't cache their hashes.

File: core/blockchain.go [hunk 3]
```diff
@@ -239,18 +239,20 @@ func (bc *BlockChain) loadLastState() error {
 	bc.currentFastBlock = bc.currentBlock
```
Comment (karalabe) [inline core/blockchain.go]: bc.currentFastBlock.Store(bc.CurrentBlock())

Comment (holiman) [reply core/blockchain.go]: Oh, nice catch!

Comment (karalabe) [reply core/blockchain.go]: It was `go vet`, not me https://travis-ci.org/ethereum/go-ethereum/jobs/345187923#L2266.

**Comment by holiman** [comment]
Ok, implemented now. The last commit needs some eyes to ensure I didn't mess something up.

**Comment by aion82** [comment]
unblok the seton if well

**Pull request merged**